### PR TITLE
Make the pionter passing to .dispatch MAYBE_NULL

### PR DESCRIPTION
**Pull request by ThinkerYzu**
This patch marks the second argument of .dispatch with
PTR_MAYBE_NULL | PTR_TO_BTF_ID | PTR_TRUSTED in
bpf_scx_is_valid_access(). The verifier  will ensures the programs
always check if the argument is NULL before reading the pointed memory.

### CODE REVIEW

File: kernel/sched/ext.c [hunk 1]
```diff
@@ -3672,6 +3672,36 @@ static bool bpf_scx_is_valid_access(int off, int size,
...
+                task_struct_btf_id = btf_find_by_name_kind(btf_vmlinux, "task_struct", BTF_KIND_STRUCT);
+                if (task_struct_btf_id < 0)
+                        return false;
+
+                info->reg_type = PTR_MAYBE_NULL | PTR_TO_BTF_ID | PTR_TRUSTED;
+                info->btf = btf_vmlinux;
+                info->btf_id = task_struct_btf_id;
+
+                return true;
+        }
```
Comment (htejun) [inline kernel/sched/ext.c:3723]: This looks great but can you please separate this out into its own function and add a bit of comment explaining what's happening?

**Review by htejun — CHANGES_REQUESTED** [review]
Looks good to me. I left a comment asking for a trivial change.

File: kernel/sched/ext.c [hunk 2]
```diff
@@ -3672,6 +3672,36 @@ static bool bpf_scx_is_valid_access(int off, int size,
...
+                return false;
+
+        t = st_ops->type;
+	member_idx = prog->expected_attach_type;
+	if (member_idx >= btf_type_vlen(t))
+                return false;
+
+	member = &btf_type_member(t)[member_idx];
+	mname = btf_name_by_offset(btf_vmlinux, member->name_off);
+        if (strcmp(mname, "dispatch") == 0 && off == sizeof(__u64) && type == BPF_READ) {
```
Comment (ByteLab-David) [inline kernel/sched/ext.c]: Could you please add a comment here explaining what we're checking here and why this corresponds to the `prev` arg of the callback?

File: kernel/sched/ext.c [hunk 3]
```diff
@@ -3672,6 +3672,36 @@ static bool bpf_scx_is_valid_access(int off, int size,
...
+                info->reg_type = PTR_MAYBE_NULL | PTR_TO_BTF_ID | PTR_TRUSTED;
+                info->btf = btf_vmlinux;
+                info->btf_id = task_struct_btf_id;
+
+                return true;
+        }
+
 	if (off < 0 || off >= sizeof(__u64) * MAX_BPF_FUNC_ARGS)
 		return false;
 	if (type != BPF_READ)
```
Comment (ByteLab-David) [inline kernel/sched/ext.c:3733]: If you bring this check to the beginning of the function, you don't have to check `BPF_READ` above

File: tools/testing/selftests/scx/Makefile [hunk 4]
```diff
@@ -175,6 +176,8 @@ testcase-targets := $(addsuffix .o,$(addprefix $(SCXOBJ_DIR)/,$(auto-test-target
 $(SCXOBJ_DIR)/runner.o: runner.c | $(SCXOBJ_DIR)
 	$(CC) $(CFLAGS) -c $< -o $@
 
+$(SCXOBJ_DIR)/maybe_null.o: $(INCLUDE_DIR)/maybe_null_fail.bpf.skel.h
+
```
Comment (ByteLab-David) [inline tools/testing/selftests/scx/Makefile:180]: This is fine -- in a follow-on change I'll update the recipe to check if there's a _fail variant of a prog and also take that as a conditional dependency

**Review by Decave — COMMENTED** [review]
Thanks a lot for taking care of this. LGTM modulo nits

File: kernel/sched/ext.c (hunk 1 above)
Comment (ThinkerYzu) [reply kernel/sched/ext.c:3723]: Sure

File: kernel/sched/ext.c (hunk 3 above)
Comment (ThinkerYzu) [reply kernel/sched/ext.c:3733]: Agree

**Review by Decave — APPROVED** [review]

**Review by htejun — APPROVED** [review]

**Pull request merged**